### ISSUE TITLE
Accuracy was not improved the custom dataset

### ISSUE THREAD

**Issue by twmht** [ouyanghaodong/DEYOv1.5#1]
I tried training in two phases using your method. In the first phase, I trained the custom dataset from scratch. In the second phase, I finetuned the custom dataset, but the mAP dropped by nearly 15% in the second phase. the validation accuracy is unstable at each epoch in the second stage, Do you have any insights on this?

**Comment by ouyanghaodong** [comment]
Could you provide more details about the training and datasets?

**Comment by ouyanghaodong** [comment]
If the encoder output before being fed into the decoder is very poor, perhaps adopting the DEYOv2 strategy of directly inputting the dense query into the decoder could resolve this issue. If this problem truly exists in the custom dataset, we will consider integrating the rank feature and greedy matching into DEYO.

**Comment by ouyanghaodong** [comment]
You can modify the ```DEYODetection``` in ```ultralytics/nn/modules/head.py``` to obtain the results of the encoder output.
```python
def get_decoder_output(self, feats, dbox, cls, imgsz):
	dbox = dbox.permute(0, 2, 1).contiguous().detach()
	cls = cls.permute(0, 2, 1).contiguous()

	bs = feats.shape[0]
	topk_ind = torch.topk(cls.max(-1).values, self.num_queries, dim=1).indices.view(-1)
	batch_ind = torch.arange(end=bs, dtype=topk_ind.dtype).unsqueeze(-1).repeat(1, self.num_queries).view(-1)
	embed = feats[batch_ind, topk_ind].view(bs, self.num_queries, -1)
	embed = self.norm(embed)

	dec_bboxes = dbox[batch_ind, topk_ind].view(bs, self.num_queries, -1)
	refer_bbox = dec_bboxes / torch.tensor(imgsz, device=dbox.device)[[1, 0, 1, 0]]

	dec_scores = self.decoder(
		embed,
		refer_bbox,
		self.dec_score_head,
		self.query_pos_head,
	)

	cls = cls[batch_ind, topk_ind].view(bs, self.num_queries, -1)

	return dec_bboxes, dec_scores, topk_ind, cls

def forward(self, x, imgsz=None):
	if self.stride[0] == 0:
		return super().forward(x)
	feats = self.get_encoder_input(x)
	preds, dbox, cls = self.generate_anchors(x)
	dec_bboxes, dec_scores, topk_ind, cls = self.get_decoder_output(feats, dbox, cls, imgsz)
	x = preds, dec_scores, topk_ind
	if self.training:
		return x
	y = torch.cat((dec_bboxes, cls.sigmoid()), -1)
	return y if self.export else (y, x)
```

**Comment by twmht** [comment]
@ouyanghaodong 

the datasets are private. 

I can provide the validation curve for reference.

![image](https://github.com/ouyanghaodong/DEYOv1.5/assets/1567200/c2c83701-3539-4d75-826d-c9355f84bc36)

Here is my first stage for training

```python
from ultralytics import YOLO
import torch
# Train from Scratch
model = YOLO("cfg/models/v10/yolov10n.yaml")
model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
# Use the model
model.train(data = "crocs.yaml", epochs = 500, scale = 0.5, mixup = 0, copy_paste = 0, device=[0,1,2,3], seed=133, batch=64)
```

and the second stage we load the checkpoint from first stage

```python
from ultralytics import DEYO
model = DEYO("cfg/models/deyo/deyov1.5n.yaml")
model.load("best.pt")

# Use the model
model.train(data = "crocs.yaml", epochs = 144, lr0 = 0.0001, lrf = 0.0001, weight_decay = 0.0001, optimizer = 'AdamW', warmup_epochs = 0, mosaic = 0, scale = 0.5, mixup = 0, copy_paste = 0, freeze = 23)

```
Do you have any advises to dig into that and find out the problems?

**Comment by ouyanghaodong** [comment]
Could you please tell me how many images are in your dataset?

**Comment by twmht** [comment]
@ouyanghaodong 

just about 3000 images for trainnig, the validation images are about 800 images.

**Comment by ouyanghaodong** [comment]
We have reproduced this phenomenon on our custom dataset, and it seems that the YOLO evaluation program may not be suitable for DEYO. We found a significant discrepancy between the evaluation results and those obtained using pycocotools. You may choose the best model based on val_cls_loss and use pycocotools to assess the model's performance. We found that our method can also align the accuracy of one-to-many branches very well on small datasets.

You can use the following code to evaluate the model's performance:
```python
from pycocotools.coco import COCO
from pycocotools.cocoeval import COCOeval

anno = COCO("gt.json")
pred = anno.loadRes("pred.json")
cocoEval = COCOeval(anno, pred, "bbox")

cocoEval.evaluate()
cocoEval.accumulate()
cocoEval.summarize()
```

**Comment by ouyanghaodong** [comment]
We have updated our post-processing code, and the issues have been successfully resolved. Now, you can directly utilize the evaluation tool from ultralytics to measure the performance of your model.

**Comment by chenzx2** [comment]
常规数据集上没有达到相应的精度，分析了下原因，发现是第一阶段的召回率很低，第二阶段没有发挥作用
